compiler/rustc_ast_lowering/src/item.rs

```diff
@@ -1096,9 +1096,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 // Check if this is a binding pattern, if so, we can optimize and avoid adding a
                 // `let <pat> = __argN;` statement. In this case, we do not rename the parameter.
                 let (ident, is_simple_parameter) = match parameter.pat.kind {
-                    hir::PatKind::Binding(hir::BindingAnnotation::Unannotated, _, ident, _) => {
-                        (ident, true)
-                    }
+                    hir::PatKind::Binding(_, _, ident, _) => (ident, true),
```

Comment:
Actually, I wonder if this would work if you changed it to `Unannotated | Mutable` instead; that would avoid changing the logic for `ref mut` and I don't think that's valid in function parameters anyway.
```suggestion
                    hir::PatKind::Binding(hir::BindingAnnotation::Unannotated | hir::BindingAnnotation::Mutable, _, ident, _) => (ident, true),
```

Reply:
> and I don't think that's valid in function parameters anyway

It clearly is, see the error message above. So yeah, I'm not sure how to fix this.

Reply:
The issue, it seems, is as described in the comments above:

```rust
// If `<pattern>` is a simple ident, then it is lowered to a single
// `let <pattern> = <pattern>;` statement as an optimization.
```

I don't think that this is considered valid:

```rust
let mut bad1 = mut bad1;
let ref mut bad2 = ref mut bad2;
```

Reply:
This comment a little further down the page elaborates on the reasoning:

https://github.com/rust-lang/rust/blob/c1011165e63480dabf1913e308da8b344dfa8f8a/compiler/rustc_ast_lowering/src/item.rs#L1129-L1132

It seems that not reusing the original argument's pattern will mean that the `HirId`s will no longer be densely assigned. I wonder if this is still a pressing concern that must be addressed before considering this PR for merging, the current CI failures aside.

Reply:
It seems that the tests now pass with the suggested change from @jyn514, but this is not yet ready to merge. I'm assuming that although, for example, `let ref mut foo = ref mut foo;` is invalid Rust, CI is passing because there are no tests in place to catch this. Once I add tests in place, we can check whether this change is safe or not.

Reply:
>  perhaps this change to the AST lowering logic would have other implications elsewhere in the compiler?

Right, this is why I asked for feedback from ecstatic-morse. Rustdoc doesn't really have a way to fix this on its end.